Code/Numerics/Conrec.h

```diff
@@ -200,4 +208,88 @@ inline void Contour(const double *d, size_t ilb, size_t iub, size_t jlb,
     }     /* i */
   }       /* j */
 }
+
+struct tplHash {
+  template <class T1, class T2, class T3>
+  std::size_t operator()(const std::tuple<T1, T2, T3> &p) const {
+    std::size_t res = 0;
+    boost::hash_combine(res, std::get<0>(p));
+    boost::hash_combine(res, std::get<1>(p));
+    boost::hash_combine(res, std::get<2>(p));
+    return res;
+  }
+};
+
+inline std::vector<std::pair<std::vector<RDGeom::Point2D>, double>>
+connectLineSegments(const std::vector<ConrecSegment> &segments,
+                    double coordMultiplier = 1000,
+                    double isoValMultiplier = 1e6) {
+  std::vector<std::pair<std::vector<RDGeom::Point2D>, double>> res;
+  std::unordered_map<std::tuple<int, int, long>, std::list<size_t>, tplHash>
+      endPointHashes;
+
+  auto makePointKey = [&coordMultiplier, &isoValMultiplier](const auto &pt,
+                                                            double isoVal) {
+    return std::make_tuple<int, int, long>(
+        std::lround(coordMultiplier * pt.x),
+        std::lround(coordMultiplier * pt.y),
+        std::lround(isoValMultiplier * isoVal));
+  };
+
+  // first hash all of the endpoints
+  for (auto i = 0u; i < segments.size(); ++i) {
+    const auto &seg = segments[i];
+    endPointHashes[makePointKey(seg.p1, seg.isoVal)].push_back(i);
+    endPointHashes[makePointKey(seg.p2, seg.isoVal)].push_back(i);
+  }
+
+  boost::dynamic_bitset<> segmentsDone(segments.size());
+  // a candidate end point hasn't been used already.
+  auto isCandidate = [&segmentsDone](const auto &pr) {
+    return !pr.second.empty() && !segmentsDone[pr.second.front()];
+  };
+  auto singlePoint =
+      std::find_if(endPointHashes.begin(), endPointHashes.end(), isCandidate);
+  while (singlePoint != endPointHashes.end()) {
+    auto segId = singlePoint->second.front();
+    auto currKey = singlePoint->first;
+    auto currVal = segments[segId].isoVal;
+    std::vector<RDGeom::Point2D> contour;
+    while (1) {
```

Comment:
I would feel better if this were

```
while (!segs.empty()) {
... contour code...
if (segs.empty()) {
  // we're at the end, push on the last point
  ...
  // remove break
} else {
  segId = segs.front();
 curry = endPtKey;
}
```

while (1) gives me the heebie-jeebies :)

Reply:
I generally share that opinion, but the loop is simple enough that it's clear that it doesn't run forever and moving away from it just complicates things.
We can come back to it at some point if we revisit this code.